example/advection/advection_package.cpp

```diff
@@ -279,9 +270,9 @@ void PreFill(MeshBlockData<Real> *rc) {
     // packing in principle unnecessary/convoluted here and just done for demonstration
     PackIndexMap imap;
     std::vector<std::string> vars({"advected", "one_minus_advected"});
-    const auto &v = rc->PackVariables(vars, imap);
-    const int in = imap["advected"].first;
-    const int out = imap["one_minus_advected"].first;
+    const auto &v = rc->PackVariables(vars, false, &imap);
```

Comment:
Good point about the coarse flag. I've added overloads for `PackVariables` that omit the coarse flag (if we put it at the end, then one would have to specify `PackVariables(..., nullptr, nullptr, true)` to pass true coarse flag.

I'm wondering if it would make more sense to just store the `PackIndexMap` and string keys in the pack itself, then we wouldn't have any function output parameters (references or pointers). I think performance shouldn't be impacted because we are constructing and filling the map and keys anyway, regardless whether the user requests them or not. And since packs are cached and returned by const ref, it doesn't cost anything extra to access the map and key from a pack reference.

What do you think?

Reply:
> I'm wondering if it would make more sense to just store the PackIndexMap and string keys in the pack itself, then we wouldn't have any function output parameters (references or pointers). I think performance shouldn't be impacted because we are constructing and filling the map and keys anyway, regardless whether the user requests them or not. And since packs are cached and returned by const ref, it doesn't cost anything extra to access the map and key from a pack reference.

If it works, that could be pretty nice... but does that work? Given that the `PackindexMap` is host-only? I know @AndrewGaspar suggested returning a tuple containing the pack and the map, which would work, but feels cumbersome too, because it requires manually unpacking the tuple.

Reply:
Hmm, good point. Let me play with it a bit.

Reply:
I've updated the interface. The `PackVariables` functions now return a const ref to the cached `VarMetaPack` struct (and similarly for flux-var packs). This struct contain the same pack as before, the `PackIndexMap` and a const pointer to the key (so we don't store a duplicate copy of the key). Please take a look at this commit and let me know what you think: https://github.com/lanl/parthenon/pull/535/commits/f4ec2e38dd4f3f237bbfa53a3821faecfce413e1

I think it makes the interface much more natural and removes even  more boilerplate. Getting pack and map now looks like this:
```c++
const auto &mp = rc.PackVariables({"foo", "bar"});
const auto &v = mp.pack;
const auto &imap = mp.map;
```
Of course, one could just use `mp.map` instead of `imap` and save the line to make a convenience ref. Same for `v`.

Reply:
I don't hate it. I do worry that it's a very API breaking change... so I'd like to hear from downstream codes too about this. @pgrete @jdolence do you have thoughts?

Actually I wonder if at the very highest level, i.e., the mesh data and meshblock data, we could also provide the old style overloads too. That would still be a simplification through most of the call stack. We could provide a deprecation warning if we feel that's the way to go. What do you think?

Reply:
It's definitely an API breaking change, but so is this whole PR :)

It would definitely be good to get input from the downstream codes. I think we should reach a consensus on what the interface should be and then just go with that (without any deprecation warnings). And if the consensus is to keep all the overloads with references, I can live with that.

Reply:
I don't have a strong opinion here.
It seems that "one large state vector" approach makes the usage of the entire interface very slim.
For example, this were all the changes required in `AthenaPK` to make use of this branch: https://gitlab.com/theias/hpc/jmstone/athena-parthenon/athenapk/-/commit/2eba46d2921479e1cfad5c75326ab511c52101d3
So given that there are more voice on the LANL side, my vote is for "you decide that internally" and we'll adapt.

Reply:
Internally, I think we need a tie-breaking vote. Any opinions from the LANL side?

Reply:
I vote not to introduce a breaking change if it doesn't resolve any actual issue.

Reply:
Alright, sounds like the majority vote is to keep the old packing interface. I'll undo these changes.